src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,976 @@ mod v1 {
+
+            if s.exists() {
+                return Ok(s);
+            }
+            for cf_file in &mut s.cf_files {
+                if cf_file.size == 0 {
+                    continue;
+                }
+                let f =
+                    try!(OpenOptions::new().write(true).create_new(true).open(&cf_file.tmp_path));
+                cf_file.file = Some(f);
+                cf_file.write_digest = Some(Digest::new(crc32::IEEE));
+            }
+            let f = try!(OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(&s.meta_file.tmp_path));
+            s.meta_file.file = Some(f);
+            Ok(s)
+        }
+
+        pub fn new_for_applying<T: Into<PathBuf>>(dir: T,
+                                                  key: &SnapKey,
+                                                  size_track: Arc<RwLock<u64>>)
+                                                  -> RaftStoreResult<Snap> {
+            let s = try!(Snap::new(dir, key, size_track, false, false));
+            Ok(s)
+        }
+
+        fn init_for_building(&mut self, snap: &DbSnapshot) -> RaftStoreResult<()> {
+            if self.exists() {
+                return Ok(());
+            }
+            let env_opt = EnvOptions::new();
+            for cf_file in &mut self.cf_files {
+                if plain_file_used(cf_file.cf) {
+                    let f = try!(OpenOptions::new()
+                        .write(true)
+                        .create(true)
+                        .truncate(true)
+                        .open(&cf_file.tmp_path));
+                    cf_file.file = Some(f);
+                } else {
+                    // initialize sst file writer
+                    let handle = try!(snap.cf_handle(&cf_file.cf));
+                    let io_options = snap.get_db().get_options_cf(handle);
+                    let mut writer = SstFileWriter::new(&env_opt, &io_options, handle);
+                    box_try!(writer.open(cf_file.tmp_path.as_path().to_str().unwrap()));
+                    cf_file.sst_writer = Some(writer);
+                }
+            }
+            let file = try!(OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(&self.meta_file.tmp_path));
+            self.meta_file.file = Some(file);
+            Ok(())
+        }
+
+        fn read_snapshot_meta(&mut self) -> RaftStoreResult<SnapshotMeta> {
+            let size = try!(get_file_size(&self.meta_file.path));
+            let mut file = try!(File::open(&self.meta_file.path));
+            let mut buf = Vec::with_capacity(size as usize);
+            try!(file.read_to_end(&mut buf));
+            let mut snapshot_meta = SnapshotMeta::new();
+            try!(snapshot_meta.merge_from_bytes(&buf));
+            Ok(snapshot_meta)
+        }
+
+        fn set_snapshot_meta(&mut self, snapshot_meta: SnapshotMeta) -> RaftStoreResult<()> {
```

Comment:
I think it should be checked if the cfs in snapshot_meta is the same as `SNAPSHOT_CFS`.